### PR TITLE
check for datasets, only delete samples if none attached

### PR DESCRIPTION
**Pull request by wow-such-code**

### CODE REVIEW

**Comment by Steffengreiner** [comment]
Note: I shamelessly adjusted @wow-such-code PR with the logic necessary to edit and delete batches in our db and openbis since it's interlinked

File: project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/sample/SampleRepositoryImpl.java [hunk 1]
```diff
@@ -72,6 +76,37 @@ public Result<Collection<Sample>, ResponseCode> addAll(Project project,
     return Result.fromValue(samples);
   }
 
+  @Transactional
+  @Override
+  public Result<Collection<Sample>, ResponseCode> deleteAll(Project project,
+      Collection<Sample> samples) {
+    Collection<SampleId> sampleIds = new ArrayList<>();
+    samples.forEach(sample -> sampleIds.add(sample.sampleId()));
```
Comment (KochTobi) [inline project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/sample/SampleRepositoryImpl.java:84]: You should not modify a list in a foreach loop.
Please replace by an assignment.
```java
samples.stream().map(Sample::sampleId).toList();
```

File: project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/sample/SampleRepositoryImpl.java [hunk 2]
```diff
@@ -87,4 +122,54 @@ public Result<Collection<Sample>, SampleInformationService.ResponseCode> findSam
...
+    return Result.fromValue(samples);
+  }
+
+  @Transactional
+  @Override
+  public Result<Collection<Sample>, ResponseCode> updateAll(Project project,
+      Collection<Sample> updatedSamples) {
+    Collection<SampleId> sampleIds = new ArrayList<>();
+    updatedSamples.forEach(sample -> sampleIds.add(sample.sampleId()));
+    String commaSeperatedSampleIds = sampleIds.stream().map(Object::toString).collect(
```
Comment (KochTobi) [inline project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/sample/SampleRepositoryImpl.java:146]: use `map`  and `collect` instead.

File: project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/sample/SampleRepositoryImpl.java [hunk 3]
```diff
@@ -87,4 +122,54 @@ public Result<Collection<Sample>, SampleInformationService.ResponseCode> findSam
...
+        Collectors.joining(", "));
+    try {
+      this.qbicSampleRepository.saveAll(updatedSamples);
```
Comment (KochTobi) [inline project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/sample/SampleRepositoryImpl.java:149]: here, a result object makes sense. When the save action fails, you will likely try to fix the issue and try again or request further information,...

File: project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/sample/SampleRepositoryImpl.java [hunk 4]
```diff
@@ -87,4 +122,54 @@ public Result<Collection<Sample>, SampleInformationService.ResponseCode> findSam
...
+    } catch (Exception e) {
```
Comment (KochTobi) [inline project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/sample/SampleRepositoryImpl.java:150]: Please do not catch `Exception`

File: project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/sample/SampleRepositoryImpl.java [hunk 5]
```diff
@@ -87,4 +122,54 @@ public Result<Collection<Sample>, SampleInformationService.ResponseCode> findSam
...
+      log.error("The samples:" + commaSeperatedSampleIds + "could not be updated", e);
+      rollBackCurrentTransaction();
+      return Result.fromError(ResponseCode.UPDATE_FAILED);
+    }
+    try {
+      sampleDataRepo.updateAll(updatedSamples);
+    } catch (Exception e) {
```
Comment (KochTobi) [inline project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/sample/SampleRepositoryImpl.java:157]: please do not catch `Exception`

File: project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/sample/openbis/OpenbisConnector.java [hunk 6]
```diff
@@ -152,17 +155,6 @@ private List<ch.ethz.sis.openbis.generic.asapi.v3.dto.project.Project> searchPro
...
-    SampleSearchCriteria criteria = new SampleSearchCriteria();
-    criteria.withCode().thatEquals(code);
-
-    SampleFetchOptions options = new SampleFetchOptions();
-    SearchResult<Sample> searchResult =
-        openBisClient.getV3().searchSamples(openBisClient.getSessionToken(), criteria, options);
-
-    return searchResult.getObjects();
-  }
-
```
Comment (KochTobi) [inline project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/sample/openbis/OpenbisConnector.java:165]: why was this deleted?

File: project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/sample/QbicSampleDataRepo.java [hunk 7]
```diff
@@ -26,20 +27,13 @@ public interface QbicSampleDataRepo {
...
-   * Deletes a sample with the provided code from persistence.
+   * Removes the provided samples from persistence
    *
-   * @param sampleCode the {@link SampleCode} of the sample to delete
-   * @since 1.0.0
+   * @param projectCode the {@link ProjectCode} of the project these samples belong to
+   * @param samples     the {@link Sample} to be removed from the data repository
    */
-  void delete(SampleCode sampleCode);
+  void deleteAll(ProjectCode projectCode, Collection<Sample> samples);
```
[the remark this answers is not in the document]
Comment (wow-such-code) [reply project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/sample/QbicSampleDataRepo.java:35]: Normally yes, but not with the state of the API, unfortunately. We need to fetch experiment samples and that needs the project code. Theoretically we could get the project code from one of the samples, but that seemed even worse to me.

File: project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/sample/openbis/OpenbisConnector.java [hunk 8]
```diff
@@ -266,27 +258,104 @@ private void createOpenbisSamples(List<SampleCreation> samplesToRegister) {
...
+  public void deleteAll(ProjectCode projectCode,
+      Collection<life.qbic.projectmanagement.domain.model.sample.Sample> samples) {
+
+    Set<String> sampleCodesToDelete = samples.stream().map(sample -> sample.sampleCode().code())
+        .collect(Collectors.toSet());
+
+    //Fetch samples with potential data - sample search is not working, sorry you have to see this
+    ExperimentFetchOptions fetchOptions = new ExperimentFetchOptions();
+    fetchOptions.withSamplesUsing(fetchSamplesCompletely());
+    for (Experiment experiment : searchExperimentsByProjectCode(projectCode.value(), fetchOptions)) {
```
Comment (KochTobi) [inline project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/sample/openbis/OpenbisConnector.java:285]: Please Refactor!!!
You can use inverted if statements and earlier exit to prevent this chaining of if statements.

File: project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/sample/openbis/OpenbisConnector.java [hunk 9]
```diff
@@ -266,27 +258,104 @@ private void createOpenbisSamples(List<SampleCreation> samplesToRegister) {
...
    */
   @Override
-  public boolean sampleExists(SampleCode sampleCode) {
-    return !searchSamplesByCode(sampleCode.code()).isEmpty();
+  public void updateAll(
+      Collection<life.qbic.projectmanagement.domain.model.sample.Sample> samples)
+      throws SampleNotUpdatedException {
+    try {
+      updateOpenbisSamples(convertSamplesToSampleUpdates(samples));
+    } catch (Exception e) {
```
Comment (KochTobi) [inline project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/sample/openbis/OpenbisConnector.java:326]: do not catch `Exception`

File: project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/sample/openbis/OpenbisConnector.java [hunk 10]
```diff
@@ -266,27 +258,104 @@ private void createOpenbisSamples(List<SampleCreation> samplesToRegister) {
...
+      throw new SampleNotUpdatedException(
+          "Samples could not be updated due to " + e.getCause() + " with " + e.getMessage());
+    }
+  }
+
+  private List<SampleUpdate> convertSamplesToSampleUpdates(
+      Collection<life.qbic.projectmanagement.domain.model.sample.Sample> updatedSamples) {
+    List<SampleUpdate> updatedSamplesList = new ArrayList<>();
+    try {
+      updatedSamples.forEach(sample -> {
```
Comment (KochTobi) [inline project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/sample/openbis/OpenbisConnector.java:336]: try to avoid multi line, compley functions as lambda. 
Possible solutions:
* stream over the list twice, each operation doing a separate concern
* extract a method
* use map more and filter and other stream operations instead of a uber-Foreach

File: project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/sample/openbis/OpenbisConnector.java [hunk 11]
```diff
@@ -266,27 +258,104 @@ private void createOpenbisSamples(List<SampleCreation> samplesToRegister) {
...
+
+        String openBisSampleType = retrieveOpenBisAnalyteCode(analyteValue).or(
+            () -> analyteMapper.mapFrom(analyteValue)).orElse(DEFAULT_ANALYTE_TYPE);
+        sampleUpdate.setProperty("Q_SAMPLE_TYPE", openBisSampleType);
+        if (openBisSampleType.equals(DEFAULT_ANALYTE_TYPE)) {
+          logger("No mapping was found for " + analyteValue + " when updating sample.");
+          logger("Using default value and adding " + analyteValue + " to Q_DETAILED_ANALYTE_TYPE.");
+          sampleUpdate.setProperty("Q_DETAILED_ANALYTE_TYPE", analyteValue);
+        }
+        updatedSamplesList.add(sampleUpdate);
```
Comment (KochTobi) [inline project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/sample/openbis/OpenbisConnector.java:353]: use collect instead of modifying a collection in a consumer

File: project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/sample/openbis/OpenbisConnector.java [hunk 12]
```diff
@@ -266,27 +258,104 @@ private void createOpenbisSamples(List<SampleCreation> samplesToRegister) {
...
+      });
+      return updatedSamplesList;
+    } catch (Exception e) {
```
Comment (KochTobi) [inline project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/sample/openbis/OpenbisConnector.java:356]: please do  not catch Exception

File: project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/sample/openbis/OpenbisConnector.java [hunk 13]
```diff
@@ -266,27 +258,104 @@ private void createOpenbisSamples(List<SampleCreation> samplesToRegister) {
...
+      throw e;
```
Comment (KochTobi) [inline project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/sample/openbis/OpenbisConnector.java:357]: why catch-rethrow ?

File: project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/sample/openbis/OpenbisConnector.java [hunk 14]
```diff
@@ -388,4 +457,18 @@ static class ConnectionException extends RuntimeException {
   static class MappingNotFoundException extends RuntimeException {
 
   }
+
+  public static class SampleNotDeletedException extends RuntimeException {
```
Comment (KochTobi) [inline project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/sample/openbis/OpenbisConnector.java:461]: do you need these exceptions?

File: project-management/src/main/java/life/qbic/projectmanagement/application/DeletionService.java [hunk 15]
```diff
@@ -69,9 +90,58 @@ public Result<ExperimentId, ResponseCode> deleteAllExperimentalGroups(Experiment
...
+        ignored -> {
+          throw new ApplicationException(ErrorCode.DATA_ATTACHED_TO_SAMPLES,
+              ErrorParameters.of(batchId));
+        });
+    if (sampleDeletionResult.isError()) {
+      log.debug("Samples for batch %s could not be deleted due to %s".formatted(batchId,
+          sampleDeletionResult.getError()));
+      return Result.fromError(ResponseCode.BATCH_DELETION_FAILED);
+    }
+    return Result.fromValue(batchId);
```
Comment (KochTobi) [inline project-management/src/main/java/life/qbic/projectmanagement/application/DeletionService.java:117]: this method looks very complex. Is there anywhere where you could simplify?

File: project-management/src/main/java/life/qbic/projectmanagement/application/DeletionService.java [hunk 16]
```diff
@@ -69,9 +90,58 @@ public Result<ExperimentId, ResponseCode> deleteAllExperimentalGroups(Experiment
...
+    var result = sampleDomainService.deleteSamples(project.get(),
+        samplesCollection.stream().toList());
+    if (result.isError() && result.getError()
+        .equals(SampleDomainService.ResponseCode.DATA_ATTACHED_TO_SAMPLES)) {
+      return Result.fromError(ResponseCode.DATA_ATTACHED_TO_SAMPLES);
+    }
+    if (result.isError()) {
+      return Result.fromError(ResponseCode.SAMPLE_DELETION_FAILED);
+    }
+    return Result.fromValue(result.getValue());
```
Comment (KochTobi) [inline project-management/src/main/java/life/qbic/projectmanagement/application/DeletionService.java:138]: `onError` or `mapError` could be helpful here

File: project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/sample/openbis/OpenbisConnector.java (hunk 6 above)
Comment (wow-such-code) [reply project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/sample/openbis/OpenbisConnector.java:165]: because it does not work, but throws an exception with our mix of API versions

File: project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/sample/QbicSampleDataRepo.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (wow-such-code) [reply project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/sample/QbicSampleDataRepo.java:35]: to clarify: deletion itself is possible, but checking for datasets is not.

File: project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/sample/SampleRepositoryImpl.java (hunk 1 above)
Comment (wow-such-code) [reply project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/sample/SampleRepositoryImpl.java]: changed it anyway :)

**Review by KochTobi — APPROVED** [review]
LGTM

Please note: the update of samples in openbis does not work currently
